grails-app/services/org/pih/warehouse/inventory/CycleCountProductAvailabilityService.groovy

```diff
@@ -0,0 +1,139 @@
+     *
+     * @return true if one or more cycle count items were added to the count or had their QoH modified.
+     */
+    boolean refreshProductAvailability(CycleCount cycleCount) {
+        boolean itemsHaveChanged = false
+
+        // We only want to refresh the items of the most recent count. We do this because previous counts are already
+        // completed, so their data should be considered static.
+        Set<CycleCountItem> cycleCountItems = cycleCount.getItemsOfMostRecentCount()
+        int currentCountIndex = cycleCountItems.first().countIndex
+        Location facility = cycleCount.facility
+
+        // Used by newly created items. We know that the frontend only supports a single assignee and date counted
+        // for all items of a count, so for convenience we set the values of any new items to match.
+        User assigneeForNewItems = cycleCountItems.find{ it.assignee }?.assignee
+        Date dateCountedForNewItems = cycleCountItems.find{ it.dateCounted }?.dateCounted ?: new Date()
+
+        // This logic is shared between count and recount flows so we need to check which we're in.
+        CycleCountItemStatus statusForNewItems = cycleCount.status.isCounting() ?
+                CycleCountItemStatus.COUNTING : CycleCountItemStatus.INVESTIGATING
+
+        // The refresh is based on the current product availability in all [bin location + lot] combinations for all
+        // the products of the count. We do this because new items might have been created since the count started.
+        List<AvailableItem> availableItems = getAvailableItems(facility, cycleCount.products)
+        for (AvailableItem availableItem : availableItems) {
+            boolean itemHasChanged = updateItem(cycleCount, cycleCountItems, availableItem, facility, currentCountIndex,
+                    assigneeForNewItems, dateCountedForNewItems, statusForNewItems)
+
+            itemsHaveChanged = itemsHaveChanged || itemHasChanged
+        }
+
+        return itemsHaveChanged
+    }
+
+    /**
+     * Refreshes a single cycle count item (or creates it if it does not exist) based on the given product availability
+     * record.
+     *
+     * Note that any items that are not yet added to the cycle count will be added to the count here, and any existing
+     * (non-custom) items whose QoH has become 0 will be removed from the count. As such, this method WILL mutate
+     * the cycle count items.
+     *
+     * @return true if a new cycle count item was added to the count or an existing one had their QoH modified.
+     */
+    private boolean updateItem(CycleCount cycleCount, Set<CycleCountItem> currentCountItems,
+                               AvailableItem availableItem, Location facility, int currentCountIndex,
+                               User assigneeForNewItems, Date dateCountedForNewItems,
+                               CycleCountItemStatus statusForNewItems) {
```

Comment:
I would suggest creating a class for this amount of arguments being passed to a method. I think we once agreed that 3-4 arguments should be the maximum.

Reply:
Just create a command class something like UpdateOrCreateCycleCountItemCommand.

Reply:
You could also potentially break this into multiple methods (find or create vs detect changes vs update).

Reply:
The method that detect changes could also return data (or a command) to tell you what operation needs to be executed: UpdateCycleCountItemCommand, DeleteCycleCountItemCommand, etc.